tidy3d/components/source.py

```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
         return const * offset * oscillation * amp
 
 
-SourceTimeType = Union[GaussianPulse, ContinuousWave]
+class CustomSourceTime(Pulse):
+    """Custom source time dependence.
+
+    Note
+    ----
+    To interpolate to simulation time steps, the value provided at the nearest time is used.
+
+    Example
+    -------
+    >>> sourcetime = CustomSourceTime.from_numpy(values=np.linspace(0, 1, 100), dt=.1)
+    """
+
+    sourcetime_dataset: Optional[TimeDataset] = pydantic.Field(
+        ..., title="Source time dataset", description="Dataset for storing the custom source time."
+    )
+
+    _sourcetime_dataset_none_warning = warn_if_dataset_none("sourcetime_dataset")
+
+    @classmethod
+    def from_numpy(
+        cls, freq0: float, fwidth: float, values: ArrayComplex1D, dt: float
+    ) -> "CustomSourceTime":
+        """Create a CustomSourceTime from a numpy array.
```

Comment:
I think it needs some more explicit  mention of what time coordinate are used (assume the user isn't looking at the source) for example.

```
The time coordinates used are evenly spaced between 0 and dt * (N-1) with a step size of `dt` where N is the length of the values array.

Reply:
maybee line 342 is a good place for this.